R/futuremice.R

```diff
@@ -0,0 +1,175 @@
+#' generally benefit from much faster cluster computation if \code{type} is set to \code{type = "FORK"}.
+#' @param ... Named arguments that are passed down to function \code{\link{mice}} or
+#' \code{\link{makeCluster}}.
+#'
+#' @return A mids object as defined by \code{\link{mids-class}}
+#'
+#' @author Gerko Vink, Rianne Schouten
+#' @seealso \code{\link{parallel}}, \code{\link{parLapply}}, \code{\link{makeCluster}},
+#' \code{\link{mice}}, \code{\link{mids-class}}
+#' @references
+#' Schouten, R. and Vink, G. (2017). parlmice: faster, paraleller, micer.
+#' \url{https://www.gerkovink.com/parlMICE/Vignette_parlMICE.html}
+#'
+#' #'Van Buuren, S. (2018).
+#' \href{https://stefvanbuuren.name/fimd/parallel-computation.html}{\emph{Flexible Imputation of Missing Data. Second Edition.}}
+#' Chapman & Hall/CRC. Boca Raton, FL.
+#'
+#' @examples
+#' # 150 imputations in dataset nhanes, performed by 3 cores
+#' \dontrun{
+#' imp1 <- parlmice(data = nhanes, n.core = 3, n.imp.core = 50)
+#' # Making use of arguments in mice.
+#' imp2 <- parlmice(data = nhanes, method = "norm.nob", m = 100)
+#' imp2$method
+#' fit <- with(imp2, lm(bmi ~ hyp))
+#' pool(fit)
+#' }
+#'
+#' @export
+futuremice <- function(data, m = 5, seed = NA, parallelseed = NA, n.core = NULL,
+                       n.imp.core = NULL, use.logical = TRUE, ...) {
+  
+  # check if pacakages available
+  mice:::install.on.demand("parallelly", ...)
+  mice:::install.on.demand("furrr", ...)
+  mice:::install.on.demand("future", ...)
+  
+  # check form of data and m
+  data <- check.dataform(data)
+  m <- check.m(m)
+
+  # check if data complete
+  if (sum(is.na(data)) == 0) {
+    stop("Data has no missing values")
+  }
+  
+  # number of available cores
+  available <- parallelly::availableCores(logical = use.logical)
+  
+  # check if arguments match CPU specifications
+  if (!is.null(n.core)) {
+    if (n.core > available) {
+      stop("Number of cores specified is greater than the number of logical cores in your CPU")
+    }
+  }
+
+  # determine course of action when not all arguments specified
+  if (!is.null(n.core) & is.null(n.imp.core)) {
+    n.imp.core <- m
+    warning(paste("Number of imputations per core not specified: n.imp.core = m =", m, "has been used"))
+  }
+  if (is.null(n.core) & !is.null(n.imp.core)) {
+    n.core <- available - 1 # leave one out by default for other processes. users can override if wanted
+    warning(paste("Number of cores not specified. Based on your machine a value of n.core =", parallelly::availableCores(logical = TRUE) - 1, "is chosen"))
+  }
+  if (is.null(n.core) & is.null(n.imp.core)) {
+    specs <- match.cluster(n.core = (available - 1), m = m)
+    n.core <- specs$cores
+    n.imp.core <- specs$imps
+  }
+  if (!is.na(seed)) {
+    if (n.core > 1) {
+      warning("Be careful; the specified seed is equal for all imputations. Please consider specifying parallelseed instead.")
+    }
+  }
+  if (is.na(parallelseed)) {
```

Comment:
Add check for parallelseed. If it is not specified, there are two options that are specified below (option two is best I think), and definitely works (because I checked).